trio/_socket.py

```diff
@@ -504,6 +506,13 @@ def shutdown(self, flag):
                     "address should be a (host, port, [flowinfo, [scopeid]]) "
                     "tuple"
                 )
+        elif self._sock.family == AF_UNIX:
+            await _core.checkpoint()
+            # unwrap path-likes
+            if hasattr(address, "__fspath__"):
+                return address.__fspath__()
+            return address
```

Comment:
Oh, excellent catch.

At first I was going to suggest using [`os.fspath`](https://docs.python.org/3/library/os.html#os.fspath), but then I remembered that that's only added in 3.6. We could skip this check on 3.5, but... we probably want to support `trio.Path` objects even on 3.5. And then I looked again at trio's async file I/O code, and we actually have a second copy of this `os.fspath` emulation there.

Would you be up to refactoring this into a `trio._util.fspath` helper? Ideally [based on the sample code here](https://www.python.org/dev/peps/pep-0519/#os)? (It turns out that calling `address.__fspath__()` directly is subtly wrong, because python's magic method lookup is weird.)

Reply:
I also filed [bpo-32562](https://bugs.python.org/issue32562).

Reply:
Sure, I'll update it when I get home today.